## Load de train and test data

In [1]:
import sys
import os
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader
from torchvision import transforms

# Add the parent directory to sys.path to access mynet module
current_dir = os.getcwd()
parent_dir = os.path.dirname(current_dir)
if parent_dir not in sys.path:
    sys.path.insert(0, parent_dir)

# Importar tus componentes (después de agregar el path)
from mynet.dataset import CustomMNISTDataset
from mynet.model import ThreeLayerFullyConnectedNetwork
from mynet.main import train_and_evaluate_model

# Configurar transformaciones
transform = transforms.Compose([
    transforms.ToTensor(),
    #transforms.Normalize((0.5,), (0.5,))
])

# Obtener la ruta correcta a los datos (subir 2 niveles desde src/mynet/)
project_root = os.path.abspath(os.path.join(os.getcwd(), '../..'))
data_path = os.path.join(project_root, 'data')

print(f"Project root: {project_root}")
print(f"Data root: {data_path}")

# Cargar dataset de entrenamiento
train_dataset = CustomMNISTDataset(root=data_path, subset='train', transformation=transform)
print(f"\nlenght train: {len(train_dataset)}")
# Cargar dataset de prueba
test_dataset = CustomMNISTDataset(root=data_path, subset='test', transformation=transform)
print(f"\nlenght test: {len(test_dataset)}")

Project root: c:\T53_NewFeatures\dlcv25-assignment-1-Cesar421
Data root: c:\T53_NewFeatures\dlcv25-assignment-1-Cesar421\data

lenght train: 60000

lenght test: 10000


In [2]:
print("\n=== Probando una muestra ===")
sample_image, sample_label = train_dataset[5]
print(f"Shape de la imagen: {sample_image.shape}")
print(f"Etiqueta: {sample_label}")
print(f"Rango de valores: [{sample_image.min():.3f}, {sample_image.max():.3f}]")



=== Probando una muestra ===
Shape de la imagen: torch.Size([1, 28, 28])
Etiqueta: 2
Rango de valores: [0.000, 1.000]


## Check the model
## Load and validate 1 epcoh

In [5]:
from mynet.main import train_one_epoch
from mynet.main import evaluate_one_epoch


model = ThreeLayerFullyConnectedNetwork()
print(f"Model parameters {model.parameters()}")
for name, param in model.named_parameters():
    print(f"{name}: {param.shape}")

# Entrenar el modelo por una época usando train_one_epoch

# Definir el optimizador y la función de pérdida
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
#optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
criterion = nn.CrossEntropyLoss()
num_epochs = 10

# Crear el DataLoader para el conjunto de entrenamiento
train_loader = DataLoader(train_dataset, batch_size=500, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=64, shuffle=True) 
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model.to(device)

# Entrenar una época
average_loss = train_one_epoch(model, train_loader, criterion, optimizer, device);
print(f"average_loss: {average_loss:.4f}");
evaloneepcoh = evaluate_one_epoch(model, test_loader, criterion, device);
print(f"evaloneepcoh: {evaloneepcoh}");

trainmodel = train_and_evaluate_model(model, train_loader, test_loader, criterion, optimizer, num_epochs, device, scheduler=None)
print(f"trainmodel: {trainmodel}");

Model parameters <generator object Module.parameters at 0x0000026284D08EB0>
fc1.weight: torch.Size([32, 784])
fc1.bias: torch.Size([32])
fc2.weight: torch.Size([64, 32])
fc2.bias: torch.Size([64])
fc3.weight: torch.Size([10, 64])
fc3.bias: torch.Size([10])
average_loss: 0.9966
evaloneepcoh: (0.3562555284636795, 0.9005)
trainmodel: ([0.31932088832060496, 0.2586125465730826, 0.2263203203678131, 0.2019434817135334, 0.1840696437905232, 0.16616708518316348, 0.15491912340124447, 0.14033970578263202, 0.1307406605531772, 0.12198344903687636], [0.2660559111150207, 0.23105702525491167, 0.20354057074921905, 0.1928103644472019, 0.17588430994255527, 0.16041921082717978, 0.1618204186106943, 0.14450971413237654, 0.1345970902330936, 0.13570311944338548], [0.9255, 0.9329, 0.9396, 0.9441, 0.9487, 0.9514, 0.9512, 0.9565, 0.9594, 0.9574])


Model parameters <generator object Module.parameters at 0x0000022E0970F5F0>
fc1.weight: torch.Size([32, 784])
fc1.bias: torch.Size([32])
fc2.weight: torch.Size([64, 32])
fc2.bias: torch.Size([64])
fc3.weight: torch.Size([10, 64])
fc3.bias: torch.Size([10])
average_loss: 2.2813
evaloneepcoh: (2.256647479002643, 0.2261)

Model parameters <generator object Module.parameters at 0x0000022E0970F510>
fc1.weight: torch.Size([32, 784])
fc1.bias: torch.Size([32])
fc2.weight: torch.Size([64, 32])
fc2.bias: torch.Size([64])
fc3.weight: torch.Size([10, 64])
fc3.bias: torch.Size([10])
average_loss: 2.3043
evaloneepcoh: (2.30256241901665, 0.1355)

In [7]:
trainmodel

([0.31932088832060496,
  0.2586125465730826,
  0.2263203203678131,
  0.2019434817135334,
  0.1840696437905232,
  0.16616708518316348,
  0.15491912340124447,
  0.14033970578263202,
  0.1307406605531772,
  0.12198344903687636],
 [0.2660559111150207,
  0.23105702525491167,
  0.20354057074921905,
  0.1928103644472019,
  0.17588430994255527,
  0.16041921082717978,
  0.1618204186106943,
  0.14450971413237654,
  0.1345970902330936,
  0.13570311944338548],
 [0.9255,
  0.9329,
  0.9396,
  0.9441,
  0.9487,
  0.9514,
  0.9512,
  0.9565,
  0.9594,
  0.9574])